language/vm/vm_runtime/src/interpreter.rs

```diff
@@ -0,0 +1,696 @@
+        let lhs = self.operand_stack.pop_as::<T>()?;
+        let result = f(lhs, rhs);
+        if let Some(v) = result {
+            self.operand_stack.push(v)?;
+            Ok(())
+        } else {
+            Err(VMStatus::new(StatusCode::ARITHMETIC_ERROR))
+        }
+    }
+
+    fn binop_int<F, T>(&mut self, f: F) -> VMResult<()>
+    where
+        Option<T>: From<Value>,
+        F: FnOnce(T, T) -> Option<u64>,
+    {
+        self.binop(|lhs, rhs| f(lhs, rhs).map(Value::u64))
+    }
+
+    fn binop_bool<F, T>(&mut self, f: F) -> VMResult<()>
+    where
+        Option<T>: From<Value>,
+        F: FnOnce(T, T) -> bool,
+    {
+        self.binop(|lhs, rhs| Some(Value::bool(f(lhs, rhs))))
+    }
+
+    fn global_data_op<F>(
+        &mut self,
+        address: AccountAddress,
+        idx: StructDefinitionIndex,
+        module: &LoadedModule,
+        op: F,
+    ) -> VMResult<()>
+    where
+        F: FnOnce(&mut Self, AccessPath, StructDef) -> VMResult<()>,
+    {
+        let ap = Self::make_access_path(module, idx, address);
+        if let Some(struct_def) =
+            self.module_cache
+                .resolve_struct_def(module, idx, &self.gas_meter)?
+        {
+            op(self, ap, struct_def)
+        } else {
+            Err(VMStatus::new(StatusCode::LINKER_ERROR))
+        }
+    }
+
+    fn borrow_global(&mut self, ap: AccessPath, struct_def: StructDef) -> VMResult<()> {
+        let global_ref = self.data_view.borrow_global(&ap, struct_def)?;
+        /*
+        self.gas_meter.calculate_and_consume(
+            &instruction,
+            &self.operand_stack,
+            global_ref.size(),
+        )?;
+        */
+        self.operand_stack.push(Value::global_ref(global_ref))
+    }
+
+    fn exists(&mut self, ap: AccessPath, struct_def: StructDef) -> VMResult<()> {
+        let (exists, mem_size) = self.data_view.resource_exists(&ap, struct_def)?;
+        /*
+        self.gas_meter.calculate_and_consume(
+            &instruction,
+            &self.operand_stack,
+            mem_size,
+        )?;
+        */
+        self.operand_stack.push(Value::bool(exists))
+    }
+
+    fn move_from(&mut self, ap: AccessPath, struct_def: StructDef) -> VMResult<()> {
+        let resource = self.data_view.move_resource_from(&ap, struct_def)?;
+        /*
+        self.gas_meter.calculate_and_consume(
+            &instruction,
+            &self.operand_stack,
+            resource.size(),
+        )?;
+        */
+        self.operand_stack.push(resource)
+    }
+
+    fn move_to_sender(&mut self, ap: AccessPath, struct_def: StructDef) -> VMResult<()> {
+        let resource = self.operand_stack.pop_as::<Struct>()?;
+        /*
+        self.gas_meter.calculate_and_consume(
+            &instruction,
+            &self.operand_stack,
+            resource.size(),
+        )?;
+        */
+        self.data_view.move_resource_to(&ap, struct_def, resource)
+    }
+
+    fn make_access_path(
```

Comment:
IIRC the ModuleId should implement `Into<AccessPath>` so we won't need such conversion here.

Reply:
cool, I'll try that, I was just moving code from the txn_executor....

Reply:
it turns out it's not the module id we need here but the resource key which is what that call is doing. Maybe there is a way to plug that in better, not too sure at the moment and if you have suggestions please let me know...